use replxx instead of readline

Willaaaaaaa opened the pull request:
### What problem were solved in this pull request?

Issue Number: #558 

Problem:
As I mentioned in the issue, students cannot use the arrow keys to browse the command history in the current cli. The existing code seems to want to use `readline`(GNU Readline library), which may conflict with the existing license. So I used replxx to improve this function in the cli.

### What is changed and how it works?
#### Key changes:
- 3rd party dependency: replxx has been added as a git submodule
- `CMakeLists.txt` in `observer`
- related cpp source files
#### Basic logic:
1. The `CliCommunicator `class utilizes a `replxx::Replxx` instance to manage the interaction.
2. Every 5s, history saved to `./.miniob.history`.

### Other information
1. This PR focuses **solely** on the `observer -P cli` functionality. The **obclient** is **NOT** modified.
2. Default max_history_size = 1000.
3. Default config: when `current_time - previous_history_save_time > 5`, history_save.

nautaa commented on src/observer/net/cli_communicator.cpp:
```diff
@@ -12,118 +12,95 @@ See the Mulan PSL v2 for more details. */
 
-#include <setjmp.h>
-#include <signal.h>
+#include <string>
```
common/lang/string.h include <string>, so remove this.

nautaa commented on src/observer/net/cli_communicator.cpp:
```diff
@@ -12,118 +12,95 @@ See the Mulan PSL v2 for more details. */
-}
+static replxx::Replxx rx;
+const std::string     REPLXX_HISTORY_FILE = "./.miniob.history";
+// const int MAX_HISTORY_SIZE = 500;  // default: 1000
```
if we use the default value, remove this line?

Willaaaaaaa replied:
OK! My initial thought was that if someone wants to modify it, they can just uncomment it.

nautaa commented on src/observer/net/cli_communicator.cpp:
```diff
@@ -12,118 +12,95 @@ See the Mulan PSL v2 for more details. */
+
+  static bool is_first_call = true;
+  if (is_first_call) {
+    // rx.set_max_history_size(MAX_HISTORY_SIZE);
```
should we remove?

Willaaaaaaa replied:
I think I'll remove all the comments. After all, the default value is quiet suitable already.

nautaa reviewed (commented):
please also change readline to replxx in `src/client` `src/oblsm/client`

Willaaaaaaa commented:
> please also change readline to replxx in `src/client` `src/oblsm/client`

I'll try to change tonight, thanks for advice!🌹

nautaa commented on src/observer/CMakeLists.txt:
```diff
@@ -42,16 +42,11 @@ ADD_EXECUTABLE(observer ${MAIN_SRC})
+
+find_package(Threads REQUIRED)
+find_package(replxx REQUIRED)
+TARGET_LINK_LIBRARIES(observer_static replxx::replxx)
```
add this lib in line 33? `SET(LIBRARIES common pthread dl libevent::core libevent::pthreads JsonCpp::JsonCpp)`

Willaaaaaaa replied:
You mean move line 46-49 to line 32 so that the code structure can be neater?

nautaa replied:
> You mean move line 46-49 to line 32 so that the code structure can be neater?

yep

nautaa commented:
i have another question, in my environment(centos 7), it seem will build failed. but for ubuntu 24.04, it's ok. do you have any idea?
```
/data/buduo.jyh/public_miniob/miniob/src/observer/net/cli_communicator.cpp:39: undefined reference to `replxx::Replxx::input(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/data/buduo.jyh/public_miniob/miniob/src/observer/net/cli_communicator.cpp:58: undefined reference to `replxx::Replxx::history_add(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
../../lib/libobserver.a(cli_communicator.cpp.o): In function `read_command()':
/data/buduo.jyh/public_miniob/miniob/src/observer/net/cli_communicator.cpp:89: undefined reference to `replxx::Replxx::history_load(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/data/buduo.jyh/public_miniob/miniob/src/observer/net/cli_communicator.cpp:99: undefined reference to `replxx::Replxx::history_save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
../../lib/libobserver.a(cli_communicator.cpp.o): In function `CliCommunicator::~CliCommunicator()':
/data/buduo.jyh/public_miniob/miniob/src/observer/net/cli_communicator.cpp:166: undefined reference to `replxx::Replxx::history_save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```

Willaaaaaaa commented:
@nautaa 
Mine is Ubuntu 24.04, too.
I can only tell that this might have something to do with the linker. 

I have googled the issue you mentioned. Some says it can be solved by [adding `pthread` and `dl` to cmakelist](https://discourse.slicer.org/t/build-error-centos-7-undefined-reference-to-pthread/15142/2) as you have mentioned. Some says it is relevant to the [ABI mismatch issue](https://redmine.emweb.be/boards/2/topics/15468), see here: https://gcc.gnu.org/onlinedocs/libstdc/manual/using_dual_abi.html.
I think you can try to "add -D_GLIBCXX_USE_CXX11_ABI=0 to your compiler options to make sure you're using the old ABI". If it really works, maybe we need to change the `cmakelists.txt` file to add the option.

nautaa reviewed (commented):
can we use a class LineReader to encapsulate, because it looks like there is too much duplicate code.

nautaa commented:
> @nautaa Mine is Ubuntu 24.04, too. I can only tell that this might have something to do with the linker.
> 
> I have googled the issue you mentioned. Some says it can be solved by [adding `pthread` and `dl` to cmakelist](https://discourse.slicer.org/t/build-error-centos-7-undefined-reference-to-pthread/15142/2) as you have mentioned that we have done so. Some says it is relevant to the [ABI mismatch issue](https://redmine.emweb.be/boards/2/topics/15468), see here: https://gcc.gnu.org/onlinedocs/libstdc/manual/using_dual_abi.html. I think you can try to "add -D_GLIBCXX_USE_CXX11_ABI=0 to your compiler options to make sure you're using the old ABI". If it really works, maybe we need to change the `cmakelists.txt` file to add the option.

thanks for you advice! i will try it later

Willaaaaaaa commented:
> can we use a class LineReader to encapsulate, because it looks like there is too much duplicate code.

Yes! Exactly what I want to do last night. But I'm not sure where to put this class, so I just modified the original `readline` part to `replxx`.

Put the class in `./src/common/linereader`, `./src/common/io` or somewhere else? What do you think?

nautaa commented:
> > can we use a class LineReader to encapsulate, because it looks like there is too much duplicate code.
> 
> Yes! Exactly what I want to do last night. But I'm not sure where to put this class, so I just modified the original `readline` part to `replxx`.
> 
> Put the class in `./src/common/linereader`, `./src/common/io` or somewhere else? What do you think?

`./src/common/linereader` is ok

Willaaaaaaa commented:
@nautaa 
OMG, I was just browsing the issues in `replxx` and suddenly realized that this project has not been maintained for a long time. Sadly, it seems that the author has [passed away](https://github.com/AmokHuginnsson/replxx/issues/152).
Will this cause any potential problems? There might be no one to maintain this repo.
Sorry, my bad. I haven't realized this. And I haven't encountered such situation before, I don't know how to handle this. Maybe I should look for an alternative? Maybe the current `replxx` is enough?

nautaa commented:
> @nautaa OMG, I was just browsing the issues in `replxx` and suddenly realized that this project has not been maintained for a long time. Sadly, it seems that the author has [passed away](https://github.com/AmokHuginnsson/replxx/issues/152). Will this cause any potential problems? There might be no one to maintain this repo. Sorry, my bad. I haven't realized this. And I haven't encountered such situation before, I don't know how to handle this. Maybe I should look for an alternative? Maybe the current `replxx` is enough?


maybe [linenoise](https://github.com/antirez/linenoise) is better?  looks like there are more projects use it.

> Redis, MongoDB, Android and many other projects

If you want to introduce replxx continue, you can add a compile option. if someone has problems with replxx, he can turn off the compile option for replxx.

Willaaaaaaa commented:
## key changes
1. `LineReaderManager` class: 
	The original code has duplicate code like `my_readline()` and `is_exit_command()`. Therefore, they are combined into the `LineReaderManager` class. This class contains a member variable `LineReader`, which determines to use either `linenoise` or `replxx` according to a CMake macro(see info 2).
2. `LinenoiseReader` class: 
	A wrapper class for `linenoise` lib. Since the `linenoise` library has only two files, I placed them directly under `./src/common/linereader`.
## TODO
Since `replxx` and `linenoise` are C++ and C libs respectively, wrapping them both in a single class can cause issues such as memory leaks. In the latest version I submitted, `linenoise` **works** without problems, but **`replxx` still leaks memory**. Mainly because of some C-style string pointers. Really annoying! I'm fixing it currently, but I have some homework deadlines coming up soon, so the submission of the changes may not be fast.
## infor
1. For `replxx`, I did not remove it mainly because:
	- This is a **C++** lib based on `linenoise` which is a C lib. Therefore, it supports more functions than `linenoise`.
	- More existing key bindings
	- It supports UTF-8 (but we don't seem to need this function)
2. Use `replexx`: 
	> mkdir build && cd build

	> cmake .. -DUSEREPLXX=1 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_BUILD_TYPE=debug
	Use `linenoise`: cmake option **default** to not use `replxx`, so no seeting no replxx.
3. more info:  
	- I think this one can work with `replxx` but have dup code and NO `linenoise` support: `commit 280fbbf96310377f2a5d398e05bd581cd725dfa6`
	- **The latest one** can work with `linenoise` but having memory leak with `replxx`.

nautaa commented:
thanks for your effort, I will convert it to draft for now, you can mark it as ready to review when you're ready.

nautaa commented on src/common/linereader/linenoise.h:
```diff
@@ -0,0 +1,113 @@
+/* linenoise.h -- VERSION 1.0
```
you can put the linenoise lib in `deps/3rd`.

hnwyllmm commented on src/common/linereader/line_reader.h:
```diff
@@ -0,0 +1,59 @@
+#endif
+
+namespace common {
+class LineReaderManager
```
This class is not a `manager`. Do you think MiniobLineReader or MyLineReader or LineReaderWrapper is better? Or anything else.

Willaaaaaaa replied:
sorry! I really don't have talents in naming things, and my understanding in `manager` is not deep😥
I thought this class can **manage** to choose which lib to use. So i simply named it as `manager`, because I thought of `BufferPoolManager`.
I think all your suggestions are wonderful. I think I will choose one of them. Maybe `LineReaderWrapper` or `MiniobLineReader `🌹

Willaaaaaaa commented on CMakeLists.txt:
```diff
@@ -33,7 +33,7 @@ OPTION(STATIC_STDLIB "Link std library static or dynamic, such as libgcc, libstd
 OPTION(USE_MUSL_LIBC "Use musl libc" OFF)
 OPTION(WITH_CPPLINGS "Compile cpplings" ON)
-
+OPTION(USE_REPLXX "Use replxx instead of linenoise for line reading" OFF)
```
You mean introduce this lib directly? What about `linenoise` then? I set this option mainly because this can choose one of the two libs.
see comment above: ["If you want to introduce replxx continue, you can add a compile option. if someone has problems with replxx, he can turn off the compile option for replxx."](https://github.com/oceanbase/miniob/pull/559#issuecomment-2885773419)

hnwyllmm replied:
Hi, @nautaa why did you suggest use both `replxx` and `linenoise`? It looks more complex.

nautaa replied:
> Hi, @nautaa why did you suggest use both `replxx` and `linenoise`? It looks more complex. 

可能有一点误解，我的意思是看起来replxx 没有继续维护了，为了避免replxx 出现潜在的问题，可以为replxx 增加一个编译选项，以便在出现非预期问题时，我们可以继续使用miniob的命令行（不带任何 readline库）。同时维护两个 readline的库确实不太合理。我也建议使用一个 readline 库即可（replxx 或 linenoise) . @Willaaaaaaa @hnwyllmm

Willaaaaaaa replied:
> > Hi, @nautaa why did you suggest use both `replxx` and `linenoise`? It looks more complex.
> 
> 可能有一点误解，我的意思是看起来replxx 没有继续维护了，为了避免replxx 出现潜在的问题，可以为replxx 增加一个编译选项，以便在出现非预期问题时，我们可以继续使用miniob的命令行（不带任何 readline库）。同时维护两个 readline的库确实不太合理。我也建议使用一个 readline 库即可（replxx 或 linenoise) . @Willaaaaaaa @hnwyllmm

明白了老师！我的我的，是我理解错了，我以为要都加上，这样可以双重保险😂

Willaaaaaaa commented on src/common/CMakeLists.txt:
```diff
@@ -1,17 +1,29 @@
+)
+
+FILE(GLOB_RECURSE CPP_SRC *.cpp)
+FILE(GLOB C_SRC ${CMAKE_CURRENT_SOURCE_DIR}/linereader/linenoise.c)
```
Ah, yes! My mistake. No need indeed, as I give cmake the file path.

Willaaaaaaa commented on src/common/CMakeLists.txt:
```diff
@@ -1,17 +1,29 @@
+SET_SOURCE_FILES_PROPERTIES(
+    ${CMAKE_CURRENT_SOURCE_DIR}/linereader/linenoise.c
+    PROPERTIES LANGUAGE C
+)
```
I didn't add this `PROPERTIES LANGUAGE C` at first. But my language server informed me of errors. AI suggested me adding this so that compiler wouldn't treat it as cpp code.
I will try **not** to add this later. I think you are right, because it's my original thought too. Maybe I should restart my `clangd` or reload the window to see if it works or not.

Willaaaaaaa commented:
So, what I am about to do is:
1. Make `linereader` a submodule
Move `linenoise` from src/common/linereader, making it a git submodule.
Added build and link configuration for this submodule in the top cmakelist (`linenoise` uses `makefile`, so I don't build it by bash build.sh). Then, updated the cmake configuration of each parts to include and link correctly.
2. Fix memory issues
As I said in previous comment, fixed the memory leak/access out-of-bounds....(I really hate this. I am not sure whether I can fix this.
3. Fix your feedbacks

Due to many courses having just concluded and exams approaching, I’ll be **slower** to fix bugs.
If urgent, maybe you can set a deadline for me so that I can make some arrangements.

Thank you for all your suggestions and comments.
As a second-year uni student, I don’t have much experience with large projects, so some mistakes are inevitable. I hope you guys can be understanding, and please feel free to challenge or correct me — it's a valuable opportunity for me to improve.🌹💕🫡

Willaaaaaaa commented:
Use `replxx` only. If anything goes wrong in the future, we can change to `linenoise` or `replxx` forked by `clickhouse`.
`observer` and `obclient` have 4 ways to quit: `\q`, `exit`, `bye` and `interrupted`. But `oblsm_cli` only have 1 way to exit: `exit`, because I followed the description in `help` and the macro only defines `exit`.
All the history files are saved to `./build` and the name of the history file depends on which you use.

nautaa commented on src/common/linereader/line_reader.cpp:
```diff
@@ -0,0 +1,87 @@
+  return line;
+}
+
+bool MiniobLineReader::is_exit_command(const std::string &cmd, const std::string &history_file)
```
the second param is not need, you have save history outsize.

nautaa commented on src/common/linereader/line_reader.h:
```diff
@@ -0,0 +1,64 @@
+   * @brief Read a line from input
+   * @param prompt The prompt to display
+   * @param history_file path/to/file
+   * @return char* to input string or nullptr
```
return string

nautaa commented on src/common/linereader/line_reader.h:
```diff
@@ -0,0 +1,64 @@
+  static bool check_and_save_history(const std::string &history_file);
+
+private:
+  static LineReader reader_;
```
you can use singleton or add a variable at client class, so don't need to mark variables `static`. and you can add history_file when linereader init. so you don't need to pass `history file` everytime.

nautaa commented on src/common/linereader/line_reader.cpp:
```diff
@@ -0,0 +1,87 @@
+
+std::string MiniobLineReader::my_readline(const std::string &prompt, const std::string &history_file)
+{
+  if (is_first_call_) {
```
add init function, when init function called, do reader_ init.

Willaaaaaaa commented on CMakeLists.txt:
```diff
@@ -158,7 +157,7 @@ IF (EXISTS /usr/local/lib64)
 ENDIF ()
 
-INCLUDE_DIRECTORIES(. ${PROJECT_SOURCE_DIR}/deps ${PROJECT_SOURCE_DIR}/src /usr/local/include)
+INCLUDE_DIRECTORIES(. ${PROJECT_SOURCE_DIR}/deps ${PROJECT_SOURCE_DIR}/deps/3rd/usr/local/include ${PROJECT_SOURCE_DIR}/src /usr/local/include)
```
Yes, you are right!
I googled it and learned that `FIND_PACKAGE` will help find the header. Understood it. Thanks!

Willaaaaaaa commented on src/obclient/client.cpp:
```diff
@@ -237,11 +177,10 @@ int main(int argc, char *argv[])
   close(sockfd);
 
+  MiniobLineReader::save_history(LINE_HISTORY_FILE);
+  printf("Command history saved to: %s\n", LINE_HISTORY_FILE.c_str());
```
Nice advice! I will add header to support `LOG_INFO`.
What about `oblsm_cli`? I don't think it has a log to record. It seems that we can only print "Command history saved to:..." in the console when using `oblsm_cli`. Is this OK?

nautaa replied:
> Nice advice! I will add header to support `LOG_INFO`. What about `oblsm_cli`? I don't think there is a log to record in it. It seems that we can only print "Command history saved to:..." in the console when using `oblsm_cli`. Is this OK?

for observer,you can add LOG_INFO. 
for obclient and oblsm_cli. I think it's no need to print to stdout/stderr that `Command history saved to ...`. Most of the command line tools I've used don't have this prompt.

Willaaaaaaa replied:
Thanks💕!
Your suggestion makes up for my lack of experience. I will correct this after you give me the latest version for review, that is, add `LOG_INFO` only for `observer `and no output of save path for `obclient` and `oblsm_cli`.🌹

nautaa commented on src/obclient/client.cpp:
```diff
@@ -179,8 +181,7 @@ int main(int argc, char *argv[])
 
-  MiniobLineReader::save_history(LINE_HISTORY_FILE);
-  printf("Command history saved to: %s\n", LINE_HISTORY_FILE.c_str());
+  LOG_INFO("Command history saved to: %s", LINE_HISTORY_FILE.c_str());
```
my mistake, obclient has no corresponding log file. you can remove this line.

nautaa reviewed (approved):
LGTM

Willaaaaaaa commented:
Thanks a lot!! It's my pleasure to contribute to `miniob`.
Really appreciate you guys' review and suggestions!! It's a valuable learning opportunity for me!🫡🥰🤩
